python/imageutils.py

```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
+        with open(legend, 'r')  as legend:
+            filename_prefix = legend.readline().strip().split('/')[-1].split('_')[0]
+            filename_ext = legend.readline().strip().split(".", 1)[1]
+            blocks_dir = legend.readline().strip().rsplit('/', 1)[0]
+            block_shape = nib.load(legend.readline().strip()).header.get_data_shape()
+
+            print "filename_prefix: {0}, filename_ext: {1}".format(filename_prefix, filename_ext)
+            print "blocks located in {0}, block's shape is {1}".format(blocks_dir, block_shape)
+
+            legend.seek(0, 0)
+            for i, l in enumerate(legend):
+                pass;
+            # n is blocks quatity in each axis. Assume all the blocks in the same size and are all cubic.
+            n = (i + 1) ** (1.0 / 3)
+
+        (y_size, z_size, x_size) = block_shape
+
+        # before reconstruction. calculate how many slices can be loaded into memory.
+
+        # mem_one_block_one_slice
+        mem_s = (y_size * z_size) * bytes_per_voxel
+        n_obs = n ** 2
+        # number of big slices can fit into the memory: 0 or more
+        n_slices = int((mem / mem_s) // n_obs)
+
+        if n_slices == 0:
+            print "mem is set too low, only one big slice is going to be loaded intto the memory"
+            n_slices = 1
+
+        with open(reconstructed, "wb+") as reconstructed:
+
+            print "writing new header..."
+            with open(self.filepath, 'rb') as f:
+                bytes_arr = f.read(self.header_size)
+                reconstructed.write(bytes_arr)
+
+            for l, x in enumerate(range(0, X_size, x_size)):
+                print("-------------{} level blocks start...-------------".format(l))
+                start_time = time()
+                for l2, i in enumerate(range(0, x_size, n_slices)):
+                    np_arr_yz = None
+                    for z in range(0, Z_size, z_size):
+                        np_arr_y = None
+                        for y in range(0, Y_size, y_size):
+                            block_name = blocks_dir + '/' + filename_prefix + "_" + str(y) + "_" + str(z) + "_" + str(
+                                x) + "." + filename_ext
+                            print "{} is loaded into memory".format(block_name)
+                            block_data = nib.load(block_name).get_data(caching='unchanged')
+                            block_data_arr = block_data[:, :, i: (i + n_slices if i + n_slices <= x_size else x_size)]
+                            np_arr_y = block_data_arr if np_arr_y is None else np.vstack((np_arr_y, block_data_arr))
+
+                        np_arr_yz = np_arr_y if np_arr_yz is None else np.hstack((np_arr_yz, np_arr_y))
+
+                    postion = self.header_size + bytes_per_voxel * (
+                        l2 * (n_slices * (Y_size * Z_size)) + l * (x_size * Y_size * Z_size))
+                    reconstructed.seek(postion, 0)
```

Comment:
you should be able to seek only once every `n_slices` while now you seek much more than that. Also, that part of the code would deserve a little bit more of comments, or better named variables to understand what's going on.